// mssql-query-builder/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = Object.setPrototypeOf ||
        ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
        function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __values = (this && this.__values) || function (o) {
    var m = typeof Symbol === "function" && o[Symbol.iterator], i = 0;
    if (m) return m.call(o);
    return {
        next: function () {
            if (o && i >= o.length) o = void 0;
            return { value: o && o[i++], done: !o };
        }
    };
};
Object.defineProperty(exports, "__esModule", { value: true });
var events_1 = require("events");
var mssql_1 = require("mssql");
var alias_1 = require("./alias");
var isOrderByClause_1 = require("./isOrderByClause");
var sql_1 = require("./sql");
var clauses = function () {
    var c = [];
    for (var _i = 0; _i < arguments.length; _i++) {
        c[_i] = arguments[_i];
    }
    return c.map(function (a) { return '(?:(?:^| )?' + a + ')?'; }).join('');
};
var objectEntries = function (obj) {
    return Object.keys(obj).map(function (key) {
        return [key, obj[key]];
    });
};
var QueryBuilder = (function (_super) {
    __extends(QueryBuilder, _super);
    function QueryBuilder(c) {
        var _this = _super.call(this) || this;
        _this._all = true;
        _this._fetch = 0;
        _this._from = null;
        _this._groupBy = [];
        _this._having = [];
        _this._input = new Map();
        _this._offset = 0;
        _this._orderBy = [];
        _this._recordSet = null;
        _this._select = new Map();
        _this._timeEnd = 0;
        _this._timeStart = 0;
        _this._top = 0;
        _this._where = [];
        _this._vars = 0;
        _this._request = sql_1.sqlRequest(c);
        return _this;
    }
    QueryBuilder.prototype.all = function (all) {
        if (all === void 0) { all = true; }
        this._all = all;
        return this;
    };
    QueryBuilder.prototype.buildQuery = function () {
        return (this._buildSelect() +
            this._buildFrom() +
            this._buildWhere() +
            this._buildGroupBy() +
            this._buildHaving() +
            this._buildOrderBy() +
            this._buildOffset() +
            ';');
    };
    QueryBuilder.prototype.distinct = function (distinct) {
        if (distinct === void 0) { distinct = true; }
        var _a, _b, e_1, _c, e_2, _d;
        if (typeof distinct === 'boolean') {
            this._all = !distinct;
            return this;
        }
        if (typeof distinct === 'string') {
            var d = this.getExpressionForAlias(distinct);
            var q = (_a = (_b = new QueryBuilder()
                .select(d + ' AS "distinct"')
                .from(this._from)).where.apply(_b, __spread(this._where)).groupBy(d)).having.apply(_a, __spread(this._having)).orderBy('"distinct"')
                .offset(this._offset)
                .fetch(this._fetch)
                .recordSet(function (rows) {
                return rows.map(function (row) {
                    return row.distinct;
                });
            });
            try {
                for (var _e = __values(this._input), _f = _e.next(); !_f.done; _f = _e.next()) {
                    var _g = __read(_f.value, 2), key = _g[0], input = _g[1];
                    q.input(input.value, key, input.type);
                }
            }
            catch (e_1_1) { e_1 = { error: e_1_1 }; }
            finally {
                try {
                    if (_f && !_f.done && (_c = _e.return)) _c.call(_e);
                }
                finally { if (e_1) throw e_1.error; }
            }
            return q.execute();
        }
        var results = [];
        try {
            for (var distinct_1 = __values(distinct), distinct_1_1 = distinct_1.next(); !distinct_1_1.done; distinct_1_1 = distinct_1.next()) {
                var d = distinct_1_1.value;
                results.push(this.distinct(d));
            }
        }
        catch (e_2_1) { e_2 = { error: e_2_1 }; }
        finally {
            try {
                if (distinct_1_1 && !distinct_1_1.done && (_d = distinct_1.return)) _d.call(distinct_1);
            }
            finally { if (e_2) throw e_2.error; }
        }
        return Promise.all(results);
    };
    QueryBuilder.prototype.execute = function (q) {
        return __awaiter(this, void 0, void 0, function () {
            var e_3, _a, query, request, _b, _c, _d, key, input, result, error_1;
            return __generator(this, function (_e) {
                switch (_e.label) {
                    case 0:
                        if (q) {
                            this.query(q);
                        }
                        query = this.buildQuery();
                        return [4, this._request];
                    case 1:
                        request = _e.sent();
                        try {
                            for (_b = __values(this._input), _c = _b.next(); !_c.done; _c = _b.next()) {
                                _d = __read(_c.value, 2), key = _d[0], input = _d[1];
                                request.input(key, input.type, input.value);
                            }
                        }
                        catch (e_3_1) { e_3 = { error: e_3_1 }; }
                        finally {
                            try {
                                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
                            }
                            finally { if (e_3) throw e_3.error; }
                        }
                        this._timeStart = Date.now();
                        _e.label = 2;
                    case 2:
                        _e.trys.push([2, 4, , 5]);
                        return [4, request.query(query)];
                    case 3:
                        result = _e.sent();
                        this._timeEnd = Date.now();
                        return [2, (typeof this._recordSet === 'number'
                                ? result.recordsets[this._recordSet]
                                : (typeof this._recordSet === 'function'
                                    ? this._recordSet(result.recordset)
                                    : result.recordset))];
                    case 4:
                        error_1 = _e.sent();
                        this._timeEnd = Date.now();
                        throw new Error(error_1.message);
                    case 5: return [2];
                }
            });
        });
    };
    QueryBuilder.prototype.fetch = function (f) {
        this._fetch = f;
        return this;
    };
    QueryBuilder.prototype.from = function (f) {
        this._from = f;
        return this;
    };
    QueryBuilder.prototype.getExpressionForAlias = function (alias) {
        return this._select.get(alias) || alias;
    };
    QueryBuilder.prototype.groupBy = function () {
        var groupBy = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            groupBy[_i] = arguments[_i];
        }
        this._groupBy = this._groupBy.concat(groupBy);
        return this;
    };
    QueryBuilder.prototype.having = function () {
        var having = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            having[_i] = arguments[_i];
        }
        this._having = this._having.concat(having);
        return this;
    };
    QueryBuilder.prototype.input = function (value, name, type) {
        if (name === undefined) {
            var n = '__QB_INPUT_' + ++this._vars + '__';
            this.input(value, n);
            return '@' + n;
        }
        if (type === undefined) {
            if (typeof value === 'boolean') {
                return this.inputBoolean(value, name);
            }
            if (typeof value === 'number') {
                return this.inputNumber(value, name);
            }
            if (typeof value === 'string') {
                return this.inputString(value, name);
            }
            if (value instanceof Buffer) {
                return this.inputBuffer(value, name);
            }
            if (value instanceof Date) {
                return this.inputDate(value, name);
            }
            throw new Error('Only boolean, number, string, Buffer, and Date inputs are allowed.');
        }
        this._input.set(name, { type: type, value: value });
        return this;
    };
    QueryBuilder.prototype.inputBit = function (value, name) {
        return this.input(value, name, mssql_1.Bit);
    };
    QueryBuilder.prototype.inputBoolean = function (value, name) {
        return this.inputBit(value, name);
    };
    QueryBuilder.prototype.inputBuffer = function (value, name) {
        return this.inputVarBinary(value, name);
    };
    QueryBuilder.prototype.inputDate = function (value, name) {
        return this.inputDateTime(value, name);
    };
    QueryBuilder.prototype.inputDateTime = function (value, name) {
        return this.input(value, name, mssql_1.DateTime);
    };
    QueryBuilder.prototype.inputInt = function (value, name) {
        return this.input(value, name, mssql_1.Int);
    };
    QueryBuilder.prototype.inputNumber = function (value, name) {
        return this.inputInt(value, name);
    };
    QueryBuilder.prototype.inputNVarChar = function (value, name) {
        return this.input(value, name, mssql_1.NVarChar);
    };
    QueryBuilder.prototype.inputString = function (value, name) {
        return this.inputNVarChar(value, name);
    };
    QueryBuilder.prototype.inputVarBinary = function (value, name) {
        return this.input(value, name, mssql_1.VarBinary);
    };
    QueryBuilder.prototype.offset = function (o) {
        this._offset = o;
        return this;
    };
    QueryBuilder.prototype.orderBy = function (o) {
        var e_4, _a;
        if (Array.isArray(o)) {
            try {
                for (var o_1 = __values(o), o_1_1 = o_1.next(); !o_1_1.done; o_1_1 = o_1.next()) {
                    var orderBy = o_1_1.value;
                    this.orderBy(orderBy);
                }
            }
            catch (e_4_1) { e_4 = { error: e_4_1 }; }
            finally {
                try {
                    if (o_1_1 && !o_1_1.done && (_a = o_1.return)) _a.call(o_1);
                }
                finally { if (e_4) throw e_4.error; }
            }
        }
        else if (typeof o === 'object') {
            if (isOrderByClause_1.default(o)) {
                this.orderBy(o.by + ' ' + o.order);
            }
            else {
                throw new Error('Invalid ORDER BY clause object supplied to method `orderBy`.');
            }
        }
        else {
            var newOrder_1 = o.replace(/(?: ASC| DESC)$/, '');
            var previousOrderByIndex = this._orderBy.findIndex(function (previousOrderBy) {
                return previousOrderBy.substr(0, newOrder_1.length) === newOrder_1;
            });
            if (previousOrderByIndex > -1) {
                this._orderBy.splice(previousOrderByIndex, 1);
            }
            this._orderBy.push(o);
        }
        return this;
    };
    QueryBuilder.prototype.query = function (q) {
        q = q.replace(/[\n\r\s\t]+/g, ' ');
        var query = q.match(new RegExp(clauses('SELECT(?: (ALL|DISTINCT))?(?: TOP (\\d+))? (.+?)', 'FROM (.+?)', 'WHERE (.+?)', 'GROUP BY (.+?)', 'HAVING (.+?)', 'ORDER BY (.+?)', 'OFFSET (\\d+) ROWS?(?: FETCH (?:FIRST|NEXT) (\\d+) ROWS? ONLY)?') +
            ' ?\;? ?$'));
        if (query) {
            var _a = __read(query, 11), all = _a[1], top_1 = _a[2], select = _a[3], from = _a[4], where = _a[5], groupBy = _a[6], having = _a[7], orderBy = _a[8], offset = _a[9], fetch_1 = _a[10];
            if (select) {
                switch (all) {
                    case 'ALL':
                        this.all();
                        break;
                    case 'DISTINCT':
                        this.distinct();
                        break;
                }
                if (top_1) {
                    this.top(parseInt(top_1, 10));
                }
                this.select(select);
            }
            if (from) {
                this.from(from);
            }
            if (where) {
                this.where(where);
            }
            if (groupBy) {
                this.groupBy(groupBy);
            }
            if (having) {
                this.having(having);
            }
            if (orderBy) {
                this.orderBy(orderBy);
            }
            if (offset) {
                this.offset(parseInt(offset, 10));
                if (fetch_1) {
                    this.fetch(parseInt(fetch_1, 10));
                }
            }
        }
        else {
            throw new Error('Invalid Query provided to QueryBuilder.\n' +
                '---------------------------------------\n' +
                q);
        }
        return this;
    };
    QueryBuilder.prototype.recordSet = function (r) {
        this._recordSet = r;
        return this;
    };
    QueryBuilder.prototype.rowCount = function () {
        var _a, _b, _c, e_5, _d;
        var q = (_a = (_b = (_c = new QueryBuilder()
            .select('COUNT(*) AS "rows"')
            .all(this._all)
            .from(this._from)).where.apply(_c, __spread(this._where))).groupBy.apply(_b, __spread(this._groupBy))).having.apply(_a, __spread(this._having)).recordSet(function (r) {
            return r[0].rows;
        });
        try {
            for (var _e = __values(this._input), _f = _e.next(); !_f.done; _f = _e.next()) {
                var _g = __read(_f.value, 2), key = _g[0], input = _g[1];
                q.input(input.value, key, input.type);
            }
        }
        catch (e_5_1) { e_5 = { error: e_5_1 }; }
        finally {
            try {
                if (_f && !_f.done && (_d = _e.return)) _d.call(_e);
            }
            finally { if (e_5) throw e_5.error; }
        }
        return q.execute();
    };
    QueryBuilder.prototype.select = function () {
        var select = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            select[_i] = arguments[_i];
        }
        var e_6, _a, e_7, _b;
        try {
            for (var select_1 = __values(select), select_1_1 = select_1.next(); !select_1_1.done; select_1_1 = select_1.next()) {
                var column = select_1_1.value;
                var entries = objectEntries(typeof column === 'string' ?
                    alias_1.default(column) :
                    column);
                try {
                    for (var entries_1 = __values(entries), entries_1_1 = entries_1.next(); !entries_1_1.done; entries_1_1 = entries_1.next()) {
                        var _c = __read(entries_1_1.value, 2), alias = _c[0], expression = _c[1];
                        this._select.set(alias, expression);
                    }
                }
                catch (e_7_1) { e_7 = { error: e_7_1 }; }
                finally {
                    try {
                        if (entries_1_1 && !entries_1_1.done && (_b = entries_1.return)) _b.call(entries_1);
                    }
                    finally { if (e_7) throw e_7.error; }
                }
            }
        }
        catch (e_6_1) { e_6 = { error: e_6_1 }; }
        finally {
            try {
                if (select_1_1 && !select_1_1.done && (_a = select_1.return)) _a.call(select_1);
            }
            finally { if (e_6) throw e_6.error; }
        }
        ;
        return this;
    };
    QueryBuilder.prototype.time = function () {
        return this._timeEnd - this._timeStart;
    };
    QueryBuilder.prototype.top = function (n) {
        this._top = n;
        return this;
    };
    QueryBuilder.prototype.where = function () {
        var where = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            where[_i] = arguments[_i];
        }
        this._where = this._where.concat(where);
        return this;
    };
    QueryBuilder.prototype.whereIn = function (columnOrExpression, input) {
        var _this = this;
        if (!Array.isArray(input)) {
            return this.where(columnOrExpression + ' = ' + this.input(input));
        }
        return this.where(columnOrExpression + ' IN (' + input.map(function (value) { return _this.input(value); }).join(', ') + ')');
    };
    QueryBuilder.prototype._buildFrom = function () {
        if (this._from) {
            return ' FROM ' + this._from;
        }
        return '';
    };
    QueryBuilder.prototype._buildGroupBy = function () {
        var _this = this;
        if (this._groupBy.length) {
            return ' GROUP BY ' + this._groupBy.map(function (expression) { return _this.getExpressionForAlias(expression); }).join(', ');
        }
        return '';
    };
    QueryBuilder.prototype._buildHaving = function () {
        if (this._having.length) {
            return ' HAVING ' + this._having.join(' AND ');
        }
        return '';
    };
    QueryBuilder.prototype._buildOffset = function () {
        if (this._offset > 0 ||
            this._fetch > 0) {
            if (!this._orderBy.length) {
                throw new Error('ORDER BY is mandatory to use OFFSET and FETCH clause.');
            }
            if (this._top) {
                throw new Error('TOP cannot be combined with OFFSET and FETCH in the same query expression.');
            }
            return (' OFFSET ' + this._offset + ' ROWS' + (this._fetch > 0
                ? ' FETCH NEXT ' + this._fetch + ' ROWS ONLY'
                : ''));
        }
        return '';
    };
    QueryBuilder.prototype._buildOrderBy = function () {
        if (this._orderBy.length) {
            return ' ORDER BY ' + this._orderBy.join(', ');
        }
        return '';
    };
    QueryBuilder.prototype._buildSelect = function () {
        if (this._select.size) {
            return ('SELECT ' + (this._all
                ? 'ALL'
                : 'DISTINCT') + (this._top !== 0
                ? ' TOP ' + this._top
                : '') + ' ' +
                ((function (entries) {
                    var e_8, _a;
                    var e = [];
                    try {
                        for (var entries_2 = __values(entries), entries_2_1 = entries_2.next(); !entries_2_1.done; entries_2_1 = entries_2.next()) {
                            var entry = entries_2_1.value;
                            e.push(entry);
                        }
                    }
                    catch (e_8_1) { e_8 = { error: e_8_1 }; }
                    finally {
                        try {
                            if (entries_2_1 && !entries_2_1.done && (_a = entries_2.return)) _a.call(entries_2);
                        }
                        finally { if (e_8) throw e_8.error; }
                    }
                    return e;
                })(this._select.entries()))
                    .map(function (_a) {
                    var _b = __read(_a, 2), key = _b[0], value = _b[1];
                    return value + ' AS "' + key.replace(/\"/g, '\\\"') + '"';
                })
                    .join(', '));
        }
        return '';
    };
    QueryBuilder.prototype._buildWhere = function () {
        if (this._where.length) {
            return ' WHERE ' + this._where.join(' AND ');
        }
        return '';
    };
    QueryBuilder.default = QueryBuilder;
    return QueryBuilder;
}(events_1.EventEmitter));
module.exports = QueryBuilder;
